keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/MetropolisHastingsStep.java

```diff
@@ -0,0 +1,154 @@
+import io.improbable.keanu.network.LambdaSection;
+import io.improbable.keanu.network.NetworkSnapshot;
+import io.improbable.keanu.vertices.Vertex;
+import io.improbable.keanu.vertices.dbl.KeanuRandom;
+import lombok.Value;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+class MetropolisHastingsStep {
+
+    private static final double LOG_ZERO_PROBABILITY = Double.NEGATIVE_INFINITY;
+
+    private final ProposalDistribution proposalDistribution;
+    private final boolean useCacheOnRejection;
+    private final Map<Vertex, LambdaSection> affectedVerticesCache;
+    private final KeanuRandom random;
+
+    /**
+     * @param latentVertices       Vertices that are unknown/hidden variables
+     * @param proposalDistribution The proposal distribution
+     * @param useCacheOnRejection  True if caching values of the network such that recalculation isn't required
+     *                             on step rejection
+     * @param random               Source of randomness
+     */
+    MetropolisHastingsStep(List<? extends Vertex> latentVertices,
+                           ProposalDistribution proposalDistribution,
+                           boolean useCacheOnRejection,
+                           KeanuRandom random) {
+
+        this.proposalDistribution = proposalDistribution;
+        this.useCacheOnRejection = useCacheOnRejection;
+        this.random = random;
+        this.affectedVerticesCache = getVerticesAffectedByLatents(
+            latentVertices,
+            useCacheOnRejection
+        );
+    }
+
+    public StepResult step(final Set<Vertex> chosenVertices,
+                           final double logProbabilityBeforeStep) {
+        return step(chosenVertices, logProbabilityBeforeStep, 1.0);
+    }
+
+    /**
+     * @param chosenVertices           vertices to get a proposed change for
+     * @param logProbabilityBeforeStep The log of the previous state's probability
+     * @param T                        Temperature for simulated annealing. This
+     *                                 should be constant if no annealing is wanted
+     * @return the log probability of the network after either accepting or rejecting the sample
+     */
+    public StepResult step(final Set<Vertex> chosenVertices,
+                           final double logProbabilityBeforeStep,
+                           final double T) {
+
+        final double affectedVerticesLogProbOld = sumLogProbabilityOfAffected(chosenVertices, affectedVerticesCache);
+
+        NetworkSnapshot preProposalSnapshot = null;
+        if (useCacheOnRejection) {
+            preProposalSnapshot = getSnapshotOfAllAffectedVertices(chosenVertices, affectedVerticesCache);
+        }
+
+        Proposal proposal = proposalDistribution.getProposal(chosenVertices, random);
+        proposal.apply();
+        VertexValuePropagation.cascadeUpdate(chosenVertices);
```

Comment:
Because you never want the reject to cascade I thought it was a good idea to just say that cascading the proposal or rejection of was out of scope for the Proposal object.

Reply:
Good call